From 0c6a03138177bd9401cddaf589f667ace2aba1e5 Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Tue, 17 Sep 2013 12:48:43 +0300
Subject: ril: Enabling seek and Smart Card API support in Android

BZ: 47372

The following changes are ported from MCG Android 4.1.2
for enabling telephony in any android rebase activity.

1) Introduction of new RIL request for support of
   smart card API
2) Introduction of new RIL_Errno enumerations for
   seek and smart card API.

Change-Id: I11f629a9fb601726a0e5ca5dc7a6da45af5b4d27
Orig-Change-Id: I7f2bb74544d6fa06185b9f7c3f9653bef6e02993
Category: aosp improvement
Domain: Telephony-Telephony-FWK
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 include/telephony/ril.h |   18 ++++++++++++++++++
 libril/ril.cpp          |   11 +++++++++++
 libril/ril_commands.h   |    4 ++++
 3 files changed, 33 insertions(+), 0 deletions(-)

diff --git a/include/telephony/ril.h b/include/telephony/ril.h
index 2dae552..3a9f270 100644
--- a/include/telephony/ril.h
+++ b/include/telephony/ril.h
@@ -69,6 +69,7 @@ typedef enum {
     RIL_E_NETWORK_PUK_REQUIRED = 18,            /* Network Personalisation PUK required */
     RIL_E_NO_SUCH_ELEMENT = 19,
 
+    RIL_E_INVALID_PARAMETER = 20                /* SEEK for Android */
 } RIL_Errno;
 
 typedef enum {
@@ -302,6 +303,7 @@ typedef struct {
 } RIL_Dial;
 
 typedef struct {
+    int cla;
     int command;    /* one of the commands listed for TS 27.007 +CRSM*/
     int fileid;     /* EF id */
     char *path;     /* "pathid" from TS 27.007 +CRSM command.
@@ -316,6 +318,7 @@ typedef struct {
 } RIL_SIM_IO_v5;
 
 typedef struct {
+    int cla;
     int command;    /* one of the commands listed for TS 27.007 +CRSM*/
     int fileid;     /* EF id */
     char *path;     /* "pathid" from TS 27.007 +CRSM command.
@@ -3763,6 +3766,21 @@ typedef struct {
  */
 #define RIL_REQUEST_NV_RESET_CONFIG 121
 
+// "data" is a const RIL_SIM_IO *
+// "response" is a const RIL_SIM_IO_Response *
+#define RIL_REQUEST_SIM_TRANSMIT_BASIC 122
+
+// "data" is a const char * containing the AID of the applet
+// "response" is a int * containing the channel id
+#define RIL_REQUEST_SIM_OPEN_CHANNEL 123
+
+// "data" is a const int * containing the channel id
+// "response" is NULL
+#define RIL_REQUEST_SIM_CLOSE_CHANNEL 124
+
+// "data" is a const RIL_SIM_IO *
+// "response" is a const RIL_SIM_IO_Response *
+#define RIL_REQUEST_SIM_TRANSMIT_CHANNEL 125
 
 /***********************************************************************/
 
diff --git a/libril/ril.cpp b/libril/ril.cpp
index d72f29c..7b824aa 100644
--- a/libril/ril.cpp
+++ b/libril/ril.cpp
@@ -699,6 +699,7 @@ invalid:
 /**
  * Callee expects const RIL_SIM_IO *
  * Payload is:
+ *   int32_t cla
  *   int32_t command
  *   int32_t fileid
  *   String path
@@ -722,6 +723,12 @@ dispatchSIM_IO (Parcel &p, RequestInfo *pRI) {
 
     // note we only check status at the end
 
+    simIO.v6.cla = 0;
+    if (pRI->pCI->requestNumber != RIL_REQUEST_SIM_IO) {
+        status = p.readInt32(&t);
+        simIO.v6.cla = (int)t;
+    }
+
     status = p.readInt32(&t);
     simIO.v6.command = (int)t;
 
@@ -3960,6 +3967,10 @@ requestToString(int request) {
         case RIL_REQUEST_GET_CELL_INFO_LIST: return"GET_CELL_INFO_LIST";
         case RIL_REQUEST_SET_UNSOL_CELL_INFO_LIST_RATE: return"SET_UNSOL_CELL_INFO_LIST_RATE";
         case RIL_REQUEST_SET_INITIAL_ATTACH_APN: return "RIL_REQUEST_SET_INITIAL_ATTACH_APN";
+        case RIL_REQUEST_SIM_TRANSMIT_BASIC: return "SIM_TRANSMIT_BASIC";
+        case RIL_REQUEST_SIM_OPEN_CHANNEL: return "SIM_OPEN_CHANNEL";
+        case RIL_REQUEST_SIM_CLOSE_CHANNEL: return "SIM_CLOSE_CHANNEL";
+        case RIL_REQUEST_SIM_TRANSMIT_CHANNEL: return "SIM_TRANSMIT_CHANNEL";
         case RIL_REQUEST_IMS_REGISTRATION_STATE: return "IMS_REGISTRATION_STATE";
         case RIL_REQUEST_IMS_SEND_SMS: return "IMS_SEND_SMS";
         case RIL_REQUEST_SIM_TRANSMIT_APDU_BASIC: return "SIM_TRANSMIT_APDU_BASIC";
diff --git a/libril/ril_commands.h b/libril/ril_commands.h
index 1b1a4da..8f27bda 100644
--- a/libril/ril_commands.h
+++ b/libril/ril_commands.h
@@ -136,3 +136,7 @@
     {RIL_REQUEST_NV_WRITE_ITEM, dispatchNVWriteItem, responseVoid},
     {RIL_REQUEST_NV_WRITE_CDMA_PRL, dispatchRaw, responseVoid},
     {RIL_REQUEST_NV_RESET_CONFIG, dispatchInts, responseVoid},
+    {RIL_REQUEST_SIM_TRANSMIT_BASIC, dispatchSIM_IO, responseSIM_IO},
+    {RIL_REQUEST_SIM_OPEN_CHANNEL, dispatchString, responseInts},
+    {RIL_REQUEST_SIM_CLOSE_CHANNEL, dispatchInts, responseVoid},
+    {RIL_REQUEST_SIM_TRANSMIT_CHANNEL, dispatchSIM_IO, responseSIM_IO}
-- 
1.7.4.1

